lib/chef/knife/bootstrap/chef-full.erb

```diff
@@ -19,32 +27,128 @@ exists() {
+  grep "ERROR 404" /tmp/stderr 2>&1 >/dev/null
+  if test $? -eq 0 || test ! -s "$2"; then
+    return 1
+  fi
+  return 0
+}
+
+# do_curl URL FILENAME
+do_curl() {
+  echo "trying curl..."
+  curl -L "$1" > "$2"
+  # check for bad return status
+  [ $? -ne 0 ] && return 1
+  # check for bad output or empty file
+  grep "The specified key does not exist." "$2" 2>&1 >/dev/null
+  if test $? -eq 0 || test ! -s "$2"; then
+    return 1
+  fi
+  return 0
+}
+
+# do_fetch URL FILENAME
+do_fetch() {
+  echo "trying fetch..."
+  fetch -o "$2" "$1" 2>/tmp/stderr
+  # check for bad return status
+  test $? -ne 0 && return 1
+  return 0
+}
+
+# do_perl URL FILENAME
+do_perl() {
+  echo "trying perl..."
+  perl -e "use LWP::Simple; getprint($ARGV[0]);" "$1" > "$2"
+  # check for bad return status
+  test $? -ne 0 && return 1
+  # check for bad output or empty file
+  # grep "The specified key does not exist." "$2" 2>&1 >/dev/null
+  # if test $? -eq 0 || test ! -s "$2"; then
+  #   unable_to_retrieve_package
+  # fi
+  return 0
+}
+
+# do_python URL FILENAME
+do_python() {
+  echo "trying python..."
+  python -c "import sys,urllib2 ; sys.stdout.write(urllib2.urlopen(sys.argv[1]).read())" "$1" > "$2"
+  # check for bad return status
+  test $? -ne 0 && return 1
+  # check for bad output or empty file
+  #grep "The specified key does not exist." "$2" 2>&1 >/dev/null
+  #if test $? -eq 0 || test ! -s "$2"; then
+  #  unable_to_retrieve_package
+  #fi
+  return 0
+}
+
+# do_download URL FILENAME
+do_download() {
+  PATH=/opt/local/bin:/usr/local/sbin:/usr/local/bin:/usr/sfw/bin:/sbin:/bin:/usr/sbin:/usr/bin
+  export PATH
+
+  echo "downloading $1"
+  echo "  to file $2"
+
+  # we try all of these until we get success.
+  # perl, in particular may be present but LWP::Simple may not be installed
+
+  if exists wget; then
+    do_wget $1 $2 && return 0
+  fi
+
+  if exists curl; then
+    do_curl $1 $2 && return 0
+  fi
+
+  if exists fetch; then
+    do_fetch $1 $2 && return 0
+  fi
+
+  if exists perl; then
+    do_perl $1 $2 && return 0
+  fi
+
+  if exists python; then
+    do_python $1 $2 && return 0
+  fi
+
+  echo ">>>>>> wget, curl, fetch, perl or python not found on this instance."
+  return 16
+}
+
 <% if knife_config[:bootstrap_install_command] %>
   <%= knife_config[:bootstrap_install_command] %>
 <% else %>
   install_sh="<%= knife_config[:bootstrap_url] ? knife_config[:bootstrap_url] : "https://www.chef.io/chef/install.sh" %>"
-  if ! exists /usr/bin/chef-client; then
-    echo "Installing Chef Client..."
-    if exists wget; then
-      bash <(wget <%= "--proxy=on " if knife_config[:bootstrap_proxy] %> <%= knife_config[:bootstrap_wget_options] %> ${install_sh} -O -) <%= latest_current_chef_version_string %>
-    elif exists curl; then
-      bash <(curl -L <%= "--proxy \"#{knife_config[:bootstrap_proxy]}\" " if knife_config[:bootstrap_proxy] %> <%= knife_config[:bootstrap_curl_options] %> ${install_sh}) <%= latest_current_chef_version_string %>
```

Comment:
Whoops -- I put them back in the `do_wget` and `do_curl` routines now. Nice catch.